File: Furthest Building You Can Reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int i;
        priority_queue<int> q;
       for(i=0; i<heights.size()-1; i++){
           if(heights[i+1] > heights[i]){
               int jump = heights[i+1] - heights[i];
               if(bricks>=jump){
                   bricks -= jump;
                   q.push(jump);
               }else if(ladders > 0){
                   if(q.size()>0){
                       if(q.top()>jump){
                       bricks+=q.top();
                       q.pop();
                       q.push(jump);
                       bricks-= jump;
                   }
                   }
                       ladders--;
               }else 
                   break;
           }
       } 
        return i;
    }
};

File: Furthest Building You Can Reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        builds = []
        index = 0
        while index < len(heights)-1:
            if heights[index+1] > heights[index]:
                jump = heights[index+1] - heights[index]
                if bricks >= jump:
                    bricks -= jump
                    heappush(builds, -1*jump)
                elif ladders > 0:
                    if builds and -1 * builds[0] > jump:
                        bricks += (-1 * heappop(builds))
                        heappush(builds, -1 * jump)
                        bricks -= jump
                    ladders -= 1
                else:
                    break
            index += 1
        return index

        
